File: playground/test/paths.test.ts
/**
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { join } from "node:path";
import { expandProjectPath, PROJECTS_HOME, projectDirError, REPO_ROOT } from "../src/paths.js";

test("expandProjectPath: bare ~ and ~/ expand to home; relative paths resolve against the invocation dir", () => {
  const prevHome = process.env.HOME;
  const prevInit = process.env.INIT_CWD;
  process.env.HOME = "/tmp/fake-home";
  process.env.INIT_CWD = "/somewhere/else";
  try {
    assert.equal(expandProjectPath("~"), "/tmp/fake-home");
    assert.equal(expandProjectPath("~/apps/demo"), "/tmp/fake-home/apps/demo");
    assert.equal(expandProjectPath("rel/dir"), "/somewhere/else/rel/dir");
    assert.equal(expandProjectPath("/abs/dir"), "/abs/dir");
  } finally {
    if (prevHome === undefined) delete process.env.HOME;
    else process.env.HOME = prevHome;
    if (prevInit === undefined) delete process.env.INIT_CWD;
    else process.env.INIT_CWD = prevInit;
  }
});

test("projectDirError: repo interior refused, .projects subtree and outside dirs allowed", () => {
  assert.notEqual(projectDirError(join(REPO_ROOT, "playground")), null);
  assert.notEqual(projectDirError(PROJECTS_HOME), null, "the .projects root itself is not a project");
  assert.equal(projectDirError(join(PROJECTS_HOME, "my-app")), null);
  assert.equal(projectDirError("/tmp/anywhere/outside"), null);
});

File: services/aep-api/internal/platform/gitfs/paths_test.go
// Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
//
// WSO2 LLC. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package gitfs_test

import (
	"strings"
	"testing"

	"github.com/wso2/aep/aep-api/internal/platform/gitfs"
)

func TestPathDerivation(t *testing.T) {
	ref := gitfs.RepoRef{OrgID: "org-1", ProjectID: "_skills", RepoSlug: "org-skills"}
	const root = "/workspaces"

	repoDir, err := gitfs.RepoDir(root, ref)
	if err != nil || repoDir != "/workspaces/repos/org-1/_skills/org-skills" {
		t.Fatalf("RepoDir = (%q, %v)", repoDir, err)
	}
	snapsDir, _ := gitfs.SnapshotsDir(root, ref)
	if snapsDir != repoDir+"/snapshots" {
		t.Fatalf("derived snapshots dir = %q", snapsDir)
	}
	sha := strings.Repeat("ab", 20)
	snapDir, err := gitfs.SnapshotDir(root, ref, sha)
	if err != nil || snapDir != repoDir+"/snapshots/"+sha {
		t.Fatalf("SnapshotDir = (%q, %v)", snapDir, err)
	}
	orgDir, err := gitfs.OrgDir(root, "org-1")
	if err != nil || orgDir != "/workspaces/repos/org-1" {
		t.Fatalf("OrgDir = (%q, %v)", orgDir, err)
	}
	if gitfs.TrashDir(root) != "/workspaces/trash" || gitfs.TmpDir(root) != "/workspaces/tmp" ||
		gitfs.ReposDir(root) != "/workspaces/repos" {
		t.Fatal("root-level dirs derived wrong")
	}
}

func TestPathDerivationRejectsHostileSegments(t *testing.T) {
	const root = "/workspaces"
	base := gitfs.RepoRef{OrgID: "org", ProjectID: "proj", RepoSlug: "slug"}

	mutations := []func(*gitfs.RepoRef){
		func(r *gitfs.RepoRef) { r.OrgID = "" },
		func(r *gitfs.RepoRef) { r.OrgID = ".." },
		func(r *gitfs.RepoRef) { r.OrgID = "a/b" },
		func(r *gitfs.RepoRef) { r.ProjectID = "." },
		func(r *gitfs.RepoRef) { r.ProjectID = "p\\q" },
		func(r *gitfs.RepoRef) { r.RepoSlug = "s g" },
		func(r *gitfs.RepoRef) { r.RepoSlug = "../../etc" },
		func(r *gitfs.RepoRef) { r.RepoSlug = strings.Repeat("x", 201) },
	}
	for i, mutate := range mutations {
		ref := base
		mutate(&ref)
		if _, err := gitfs.RepoDir(root, ref); err == nil {
			t.Errorf("mutation %d: RepoDir accepted hostile segment %+v", i, ref)
		}
	}
	if _, err := gitfs.SnapshotDir(root, base, "main"); err == nil {
		t.Error("SnapshotDir accepted a non-sha")
	}
	if _, err := gitfs.SnapshotDir(root, base, strings.Repeat("AB", 20)); err == nil {
		t.Error("SnapshotDir accepted uppercase hex")
	}
	if _, err := gitfs.OrgDir(root, "or/g"); err == nil {
		t.Error("OrgDir accepted a separator")
	}
}
